psa-crypto/src/types/key_derivation_operation.rs

```diff
@@ -0,0 +1,265 @@
+// Copyright 2020 Contributors to the Parsec project.
+// SPDX-License-Identifier: Apache-2.0
+
+//! # PSA Operation types
+
+use super::algorithm::{Hash, KeyDerivation};
+use super::key::Id;
+#[cfg(feature = "interface")]
+use super::status::{Error, Result, Status};
+#[cfg(feature = "interface")]
+use core::convert::{From, TryFrom};
+
+/// Key derivation operation for deriving keys from existing sources
+#[derive(Debug, Clone)]
+pub struct KeyDerivationOperation<'a> {
+    /// Key derivation algorithm and inputs
+    pub alg: KeyDerivationWithInputs<'a>,
+    /// Maximum capacity of a key derivation operation
+    pub capacity: Option<usize>,
+}
+
+/// Wrapper around KeyDerivation to enforce correct `Input`s
+#[derive(Debug, Clone)]
+pub enum KeyDerivationWithInputs<'a> {
+    /// HKDF algorithm.
+    Hkdf {
+        /// A hash algorithm to use
+        hash_alg: Hash,
+        /// Salt, used in the "extract" step. It is optional; if omitted, the derivation uses an empty salt.
+        /// Typically a direct input, can also be a key of type `RawData`.
+        salt: Option<Input<'a>>,
+        /// Secret, used in the "extract" step. Must be `Input::Key(Id)` if used with `psa_key_derivation_output_key`.
```

Comment:
You write in multiple places:

> Must be `Input::Key(Id)` if used with `psa_key_derivation_output_key`

Would there be a way to force that in the type system?

Reply:
I couldn't think of a clean way of doing it. An option would be to have it as a parameter on the method to force `Input::Key(Id)` but that seems a bit janky. I can have another think about it after everything else is sorted.

Reply:
I'm not sure I follow... `KeyDerivationOperation` is only usable with `psa_key_derivation_output_key` - or is there more to come?